### ISSUE TITLE
Proposed Rule: align function arguments

### ISSUE THREAD

**Issue by Trott** [eslint/eslint#5946]
**When does this rule warn? Please describe and show example code:**

The rule warns when arguments in a multi-line function call are not aligned.

This code will warn:

```js
var result = myFunc('some long string that will cause someone to want to wrap',
  bar
  baz);
```

The following will not warn:

```js
var result = myFunc(somethingLongThatMightCauseSomeoneToWantToWrap,
                    bar
                    baz);
```

If an argument is itself multiline (for example, an inline function) then the rule does not fire for the next argument. This means this code will not warn even though the second argument (`100`) is not aligned with the first (the function):

```js
setTimeout(function () {
  // do some stuff
}, 100);
```

**Is this rule preventing an error or is it stylistic?**

Stylistic

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

It is a widespread practice that some developers will want to enable linting for. Some developers may already expect this to be possible.

**Are you willing to create the rule yourself?**

Sure. I have a first implementation here: https://github.com/Trott/io.js/blob/align-multiline-function-call/tools/eslint-rules/align-function-arguments.js

**Comment by kaicataldo** [comment]
@eslint/eslint-team Important for consideration for inclusion - this is an existing rule in JSCS: http://jscs.info/rule/validateAlignedFunctionParameters

**Comment by BYK** [comment]
I think this should be a part of `indent` as an option: `"CallExpression:arguments": { indent: true, align: true }`

**Comment by Trott** [comment]
> "CallExpression:arguments": { indent: true, align: true }

I think I understand what `indent: true` would do and what `align:true` would do. What would be the effect of setting both to `true`?

**Comment by nzakas** [comment]
I think we should get some insights from the JSCS folks around inclusion in `indent` vs. having a separate rule.

**Comment by mikesherov** [comment]
The problem with putting call expression indentation , (or array indentation, and to a lesser extent, object indentation) is that there isn't as much agreement on what is "correct" as block or even switch indentation. There is also a very high number of variables to consider here:

1. Should open parens be on same line as first argument? What about the closing parens?
2. Should the arguments line up with the open parens, 1 indent from the open parens, or 1 indent from the start of the line?
3. Is one argument per line required, or can you permit many?
4. That case with setTimeout above... Do you actually allow indentation to be ignored in the case of a multiline function expression or not? What if the arg is a multiline array? A multiline object?

Hopefully this would point us in the direction of having a separate rule. My original goal for JSCS was to have an indentation engine that could be used for multiple rules rather than stuffing all indentation into a single rule with massive configuration. I never got there but call expressions were the original reason why!

**Comment by nzakas** [comment]
@mikesherov how did you (or were you planning to) share indent preferences between rules? Tabs or spaces preferences need to be used in both, correct?

**Comment by mikesherov** [comment]
Ideally tabSize would be a top level key anyway. It's needed for autofixing mixed spaces and tabs too, I assume, and also needed to report location of errors, no?

**Comment by BYK** [comment]
I strongly disagree for splitting indentation enforcement into multiple rules because then it becomes very hard to make sure the whole file is indented properly.

**Comment by mikesherov** [comment]
@BYK depends on what you mean by split up. I authored the original indentation rule in JSCS, which I believe the ESLint one is still based on. I know the problem well too :-) When I say to split the rules up, I mean that there should be an underlying indentation engine that more than one rule (from the users perspective) can leverage. In the end, it's still a single pass, stack based approach.

For blocks and switch, it's easy to have a single rule. It just feels like we'd be managing a ton of edge cases for anything else, and we already see that in variableDeclaration configuration. If we don't mind monolithic config for indentation, that's fine too. We would then need the following:

1. Better config for variableDeclaration to control the case of one declarator that spans multiple lines:
```
var a = {
      a: b
    };

vs.

var b = {
  a: b
};
```
2. Need config options for call expression that accommodates all the following options:
a. Newline after opening parens and before closing parens?
b. Indent to parens or indent once or indent variable number of times?
c. Allow multiple arguments per line? (Although I suppose this can be a different rule)
d. Do multiline arguments ( be they arrays, objects, functions, template strings) require a multiline call expression or not?

**Comment by BYK** [comment]
@mikesherov I'm okay with the monolithic configuration but it feels like you've given this shared indentation engine with separate rules some thought so I'd like to hear your reasoning and plan of attack for that. I found the current implementation lacking many important things and missing certain lines when there are no "interesting" nodes on or around those lines in the current implementation.

That's why I started a rewrite but didn't have the time to finish: https://github.com/eslint/eslint/commit/4d2067341a1b186cda9fabb5ad631504c9db7017

Happy to collaborate on this since `indent` is a huge pain for our users and us.

**Comment by mikesherov** [comment]
I'm not sure how much ESLint had changed from JSCS, but it seems like a lot of special casing was introduced. Perhaps we can focus on a rewrite before we do call expression? Probably no longer relevant to discuss it in this issue, but we should make a milestone or tracking issue for all indentation related bugs we want to fix and track progress together. I'm happy to have someone else to share this challenge with.

**Comment by mikesherov** [comment]
In regards to it missing many important things, that's true and it was explicit design choice that limited scope and worked well... until we got to autofix and I realized every line needs analysis or some basic accounting. Did the best I could at the time!

**Comment by BYK** [comment]
@mikesherov oh, my fears have realized: I didn't mean to be accusing, sorry :) I appreciate the way it works now, it was just one step in its evolution and it served as quite well (still serving actually!).

So thanks! :)

**Comment by mikesherov** [comment]
No offense taken :-). I just didn't want you to believe that I thought the current way was complete.

**Comment by nzakas** [comment]
@mikesherov what does "top level key" mean? We don't have such a concept in ESLint.

I'd like to suggest tabling indent-related work until we get more of the JSCS compat work underway. `indent` is a huge pain (most open bugs are about it), so I'm afraid digging too deep now will distract us from more achievable short-term goals.

**Comment by gyandeeps** [comment]
Our indent rule is no longer based on JSCS rule. we migrated way back to use what nodeca had and then built on top of it. I was the one who did it.

**Comment by mikesherov** [comment]
@nzakas, for example, ESLint has parserOptions, which potentially effects the behavior of every rule. Line endings and tabSize have the potential to effect autofix and multiple rules. 

I'm fine with tabling for now, and having this rule just match the JSCS one.

**Comment by mikesherov** [comment]
@gyandeeps let's table this discussion for now but I'd like to take another look at JSCS's because for all its warts, it did a decent job at what it was trying to do.

**Comment by ilyavolodin** [comment]
@mikesherov If I remember correctly original implementation of indent rule was based (and by based I mean copy-pasted) from JSCS. But we had too many problems with it and it was later re-written completely based on [eslint-plugin-nodeca](https://github.com/nodeca/eslint-plugin-nodeca) implementation (which has been heavily modified and is what we are currently using).

**Comment by platinumazure** [comment]
@nzakas What @mikesherov is talking about with "top-level key" sounds a lot like the concept I've floated a few times about allowing ESLint users to store repository style settings as a separate top-level config item. This means rules, instead of depending each other, depend on a shared data store. The example I gave was quote-props being able to auto-fix because it could learn the correct quotes setting without actually depending on the quotes rule config. I know you're probably sick of me bringing it up, but surely you must see that it's naturally come up four times. I think it deserves further consideration by now.

I'll dig up the issue I filed and link it to this issue.

**Comment by alberto** [comment]
I myself proposed that too, but with autofixing I am not sure we need it anymore.

**Comment by nzakas** [comment]
@platinumazure yup, I remember. I was trying to clarify because I was unsure if @mikesherov thought such a thing existed already or not. And I never shut down previous conversations, they just never really went anywhere. If someone wants to open an issue with a proposal, we can talk about it again. (And I'd ask @mikesherov to share how JSCS solved this problem.)

@alberto we need it for when auto-fixing is not used. The rules still need to know what to do.

**Comment by mysticatea** [comment]
Just related issue: #6052. Though it's about function declarations.

**Comment by not-an-aardvark** [comment]
I implemented something along these lines in #7189, but only noticed that this issue existed afterwards.

The implementation in that PR checks only the *indentation* of CallExpression parameters. In my opinion, other validations (such as requiring all parameters to be on the same line) would be better as part of separate rule, rather than overloading `indent` to check things other than indentation.

**Comment by nzakas** [comment]
This is still indentation, just with a different target ending, so I think it could make sense as part of indent. 

As a separate rule, it would clash with indent.

**Comment by mikesherov** [comment]
TL;DR: JSCS had function call indentation separately because indentation was designed to not clash.

FWIW, JSCS did function args / param alignment as separate rule from generic indentation. It didn't conflict because the indentation rule was tri-state: correct, incorrect, or unimportant. Unimportant lines, when autofixed, moved with the previous important line.

The distinction is important, as you can consider function body, block, and switch indentation as dealing with scope and fairly standardized and option free. 

It's a bit more complicated and free form with function calls.

**Comment by ronjouch** [comment]
I upgraded from ESLint 3.8.0 to 3.9.1 and I *think* this new `CallExpression` option is causing me trouble. Consider this piece of code:

```javascript
  res.send({offset: 0,
            rows: reducedFacts})
```

Here I'm happy that `offset` and `rows` are aligned, because they're keys of the same object, but I get an `ESLint Error: indent: Expected indentation of 4 spaces but found 38` on the second line.

If I'm not misled and it's indeed due to `CallExpression`, is there a way for me to communicate that I'm okay with this indentation? If no and if there's nothing in ESLint supporting such a style, how can I disable the option *without* disabling the whole `indent` rule? The only valid option values are `first` or a positive integer, neither of which deactivates the option.

**Comment by platinumazure** [comment]
@ronjouch I'm sorry that ESLint is now calling this out without an easy way to configure the rule to support your style at this point. However, in order to support this case, you can use inline disable comments:

```js
  res.send({offset: 0,
            rows: reducedFacts})  // eslint-disable-line indent
```

You might need to finagle depending on exactly which line is being flagged, but that's something you can do in the short term.

**Comment by ronjouch** [comment]
@platinumazure thanks for the fast answer :) . Unfortunately, inline-disabling is not an option as this is far from being a one-timer; I have tons of such indents and it would be very invasive.

For now I'll disable the whole `inline` or will stick to 3.8.1. Are there any plans to support disabling options (as opposed to a whole rule)? Mabe a bug I could follow / subscribe to?

**Comment by platinumazure** [comment]
@ronjouch I think #7473 should work for you (the title says multiline array indentation, but multiline object indentation is mentioned there as well and hopefully we can cover both).
